### ISSUE TITLE
Improper systematic use of SQLDirectExecW

### ISSUE THREAD

**Issue by dgsb** [mkleehammer/pyodbc#141]
It seems on a python3 we systematically use SQLDirectW to execute a query on the driver.

```

        szLastFunction = "SQLExecDirect";
#if PY_MAJOR_VERSION < 3
        if (PyString_Check(pSql))
        {
            Py_BEGIN_ALLOW_THREADS
            ret = SQLExecDirect(cur->hstmt, (SQLCHAR*)PyString_AS_STRING(pSql), SQL_NTS);
            Py_END_ALLOW_THREADS
        }
        else
#endif
        {
            SQLWChar query(pSql);
            if (!query)
                return 0;
            Py_BEGIN_ALLOW_THREADS
            ret = SQLExecDirectW(cur->hstmt, query.get(), SQL_NTS);
            Py_END_ALLOW_THREADS
        }
```

Why do we systematically expect "UTF-16" input ?

This call on Linux host prevents me to connect to a netezza database. On python2 the query can be executed, but the following exception is raised when print the contents of a returned column.

```
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-17: ordinal not in range(128)
```

**Comment by mkleehammer** [comment]
In Python 3, all strings are Unicode.  The top portion is for non-Unicode strings which only exist in Python 2.  The Python 2 code can also drop through.

If the driver doesn't support Unicode at all then we have a problem.  If it does, it should allow input from both the A and W versions.

The usual problem is very few drivers even pay attention to the specification, so when I get a chance I'm going to add some configuration options to allow the user to choose the encodings.  For example, it would be common with MySQL to encode Unicode as UTF-8 but pass it to the *non-Unicode* APIs.  This is not allowed by the spec, but is how the driver is written.

In this case, however, I don't know where the error message came from.  The SQLWChar constructor doesn't call any APIs that I would expect to raise that.  Are you in a position to add some `printf` statements to be sure?

**Comment by dgsb** [comment]
In this case this is also another driver (netezza) which accepts utf-8 and only with the non-wide api version. I'm not an expert in odbc spec, but it seems it makes more sense to give utf-8 encoding strings to the non-wide API, don't you think ?
Regarding the debugging, you can give me a pathc, I would gladly help to fix this issue.

**Comment by dgsb** [comment]
@mkleehammer I have a look on a connection parameter to indicates the type of API to use.
Something like
``` 
pyodbc.connect(conn_str, ascii_api=True)
```
?

**Comment by mkleehammer** [comment]
OK, I have spent the weekend rewriting Unicode handling on the v4 branch to give us more flexibility.

Can you try this:

Python 2

    cnxn.setencoding(str, 'utf-8')
    cnxn.setencoding(unicode, 'utf-8')

Python 3

    cnxn.setencoding('utf-8')

If you have trouble connecting, you may need to add `encoding='utf-8'` to the connect function.

I'll write up the documentation soon but could really use some beta testing.  Thanks.

**Comment by mkleehammer** [comment]
I've rebased the v4 branch so I could make a smaller pull request #160.

**Comment by mkleehammer** [comment]
I believe Unicode issues are fixed in the newly released 4.0.0 version.
